### PR TITLE
Port 3.2 to .NET CORE and several bugfixes

### PR DESCRIPTION
**Pull request by scott-xu**
This PR omits ReadOnlyKernel as it's still at early stage.

### CODE REVIEW

**Comment by scott-xu** [comment]
@onovotny any thoughts?

**Comment by onovotny** [comment]
Okay, here's what I think -- 

1.  This PR has too many changed files... 545. That should not be required to add .NET Core support and makes this PR nearly impossible to review. The .NET Core bits should be split out from whatever else is in here (formatting, etc).
2.  Adding .NET Core support should be really easy here...so whatever was done seems like overkill. It looks like everything was moved to xproj but I can't really tell.

The approach I would recommend for adding .NET Core support is as follows:

1.  Convert the existing facade project to `netstandard1.0`. My [blog post](https://oren.codes/2016/06/23/portable-is-dead-long-live-netstandard/) shows how to do that.
2.  Add a new project, a csproj based PCL that targets `netstandard1.3`

Given the current directory structure, you'll need to ensure that the `project.json` file names include the project name -- so you'll have `Ninject.project.json`, `Ninject.ios.project.json`, etc... and you'll add a  new csproj called Ninject.netstandard and with a `Ninject.netstandard.project.json`. (be sure to update the intermediate object directories to prevent directory conflicts similar to the rest of the projects there).

Basically, I believe you can implement everything needed using a NETStandard 1.3 library and you'll simply "add an output" to the package.

**Comment by onovotny** [comment]
The other issue is that we need access to the existing CI server so we can disconnect it and to the NuGet account so we can update the original packages. Otherwise the only packages I can update is `Portable.Ninject` as I don't have the account info for the others.

**Comment by scott-xu** [comment]
This PR is a fresh porting from last released 3.2.
Since PCL is dead, we don't need any PCL specified csproj files nor PCL related conditional compilations.
Some highlights:
1. Omitted ReadOnlyKernel temporarily as its still at early stage.
2. One xproj with one project.json is enough. We can improve if xproj is completely obsoleted. 
3. Regarding platform, netstandard 1.5 plus net 4.5 should be enough. Most of the platform specified conditional compilation has been removed. 

发自我的 iPhone

**Comment by scott-xu** [comment]
I'll try set it up. 

发自我的 iPhone

**Comment by onovotny** [comment]
PCL is not dead...it's just evolved to be netstandard. It's the same concept. You do need the conditional compilations or this will stop working for anything less than netstandard1.5. In that case, you're excluding everything other than .net core apps and net 4.6.3 (which doesn't exist yet).

Doing what I suggest ensures that Ninject continues to work on all of the existing platforms. It's really easy to do. 

All the ifdef's are still required, just that the `netstandard1.3` version (one that works with net 4.6, uwp, and .net core) probably is closest to the desktop .net version.

**Comment by onovotny** [comment]
Take a look here: https://github.com/ninject/Ninject/tree/netstandard

It's compiling for net45 and netstandard1.3, so that should be all we need. It still needs some cleanup and ifdef removal, but I think it's a good starting point...

Thoughts?

**Comment by scott-xu** [comment]
What's your preference about csproj and xproj?

**Comment by scott-xu** [comment]
I'd prefer xproj before VS15 RTM, because it supports

- Build packages directly from the project
- Multi-targeting

**Comment by onovotny** [comment]
I guess I don't see the point in moving everything to xproj when we'll be getting csproj back in a few months. Besides, the WinRT bits need hacks to compile with xproj. It's just easier to not bother for now.

Once the fall hits and csproj can multi-target, then there can be some big cleanup.

The NuGet package isn't a big deal, IMO, esp since the packages already exist. It's just one more build step to call nuget pack.

**Comment by scott-xu** [comment]
Ok, let's move forward.

**Comment by scott-xu** [comment]
There are a few trick codes involved to support .netstandard 1.3. Considered about the uwp usage and to build a consolidated code base, I think .netstandard 1.5 is the way to go. I'll update the code.

**Comment by scott-xu** [comment]
Also, the upcoming tools can convert the xproj to the next version of csproj. So let's use xproj at this moment.

**Comment by onovotny** [comment]
@scott-xu I really disagree here, we cannot drop UWP. My other branch has this all done and working with both netstandard 1.4 for UWP and 1.5.
